clients/client-java/src/main/java/com/datastrato/gravitino/client/KerberosTokenProvider.java

```diff
@@ -0,0 +1,202 @@
+/*
+ * Copyright 2024 Datastrato Pvt Ltd.
+ * This software is licensed under the Apache License version 2.
+ */
+
+package com.datastrato.gravitino.client;
+
+import com.datastrato.gravitino.auth.AuthConstants;
+import com.datastrato.gravitino.auth.KerberosUtils;
+import com.google.common.base.Preconditions;
+import java.io.File;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.Base64;
+import java.util.Set;
+import java.util.concurrent.Callable;
+import javax.security.auth.kerberos.KerberosTicket;
+import javax.security.auth.login.LoginContext;
+import javax.security.auth.login.LoginException;
+import org.apache.commons.lang3.StringUtils;
+import org.ietf.jgss.GSSContext;
+import org.ietf.jgss.GSSManager;
+import org.ietf.jgss.GSSName;
+import org.ietf.jgss.Oid;
+
+/**
+ * KerberosTokenProvider will get Kerberos token using GSS context negotiation and then provide the
+ * access token for every request.
+ */
+public final class KerberosTokenProvider implements AuthDataProvider {
+
+  private final GSSManager gssManager = GSSManager.getInstance();
+  private String clientPrincipal;
+  private String keytabFile;
+  private String host = "localhost";
+  private LoginContext loginContext;
+
+  private KerberosTokenProvider() {}
+
+  /**
+   * Judge whether AuthDataProvider can provide token data.
+   *
+   * @return true if the AuthDataProvider can provide token data otherwise false.
+   */
+  @Override
+  public boolean hasTokenData() {
+    return true;
+  }
+
+  /**
+   * Acquire the data of token for authentication. The client will set the token data as HTTP header
+   * Authorization directly. So the return value should ensure token data contain the token header
+   * (eg: Bearer, Basic) if necessary.
+   *
+   * @return the token data is used for authentication.
+   */
+  @Override
+  public byte[] getTokenData() {
+    try {
+      return getTokenInternal();
+    } catch (Exception e) {
+      throw new IllegalStateException("Fail to get the Kerberos token", e);
+    }
+  }
+
+  private byte[] getTokenInternal() throws Exception {
+
+    String[] principalComponents = clientPrincipal.split("@");
+    // Gravitino server's principal must start with HTTP. This restriction follows
+    // the style of Apache Hadoop.
+    String serverPrincipal = "HTTP/" + host + "@" + principalComponents[1];
+
+    synchronized (this) {
+      if (loginContext == null) {
+        loginContext = KerberosUtils.login(clientPrincipal, keytabFile);
+      } else if (isLoginTicketExpired() && keytabFile != null) {
+        // We only support use keytab to re-login context
+        loginContext.logout();
+        loginContext = KerberosUtils.login(clientPrincipal, keytabFile);
```

Comment:
You'd do some verification locally to make sure the logic is working correctly.

Reply:
Add an extra ut.